src/test/java/com/google/devtools/build/lib/runtime/BuildSummaryStatsModuleTest.java

```diff
@@ -0,0 +1,125 @@
+  }
+
+  private ActionResultReceivedEvent createActionEvent(Duration userTime, Duration systemTime) {
+    ActionResult result = mock(ActionResult.class);
+    when(result.cumulativeCommandExecutionUserTime()).thenReturn(Optional.ofNullable(userTime));
+    when(result.cumulativeCommandExecutionSystemTime()).thenReturn(Optional.ofNullable(systemTime));
+    when(result.spawnResults()).thenReturn(ImmutableList.of());
+    return new ActionResultReceivedEvent(null, result);
+  }
+
+  private BuildCompleteEvent createBuildEvent() {
+    BuildResult buildResult = new BuildResult(1000);
+    buildResult.setStopTime(2000);
+    return new BuildCompleteEvent(buildResult);
+  }
+
+  @Test
+  public void allCpuTimesAreSummarized() throws Exception {
+    ActionResultReceivedEvent action1 = createActionEvent(Duration.ofSeconds(50), Duration.ofSeconds(20));
+    ActionResultReceivedEvent action2 = createActionEvent(Duration.ofSeconds(5), Duration.ofSeconds(2));
+    buildSummaryStatsModule.actionResultReceived(action1);
+    buildSummaryStatsModule.actionResultReceived(action2);
+    Field field1 = tBSSM.getDeclaredField("cpuTimeForBazelJvm");
+    field1.setAccessible(true);
+    field1.set(buildSummaryStatsModule, Duration.ofMillis(11000));
+    buildSummaryStatsModule.buildComplete(createBuildEvent());
+    verify(reporterMock).handle(Event.info(String.format("CPU time %.2fs (user %.2fs, system %.2fs, bazel jvm %.2fs)",88.00, 55.00, 22.00, 11.00)));
+  }
+
+  @Test
+  public void mixOfActionsWithKnownAndUnknownCpuTimesResultInUnknownTimes() throws Exception {
+    // First action with unknown values
+    ActionResultReceivedEvent action1 = createActionEvent(null, null);
+    // Followed by action with known values
+    ActionResultReceivedEvent action2 = createActionEvent(Duration.ofSeconds(50), Duration.ofSeconds(20));
+    buildSummaryStatsModule.actionResultReceived(action1);
+    buildSummaryStatsModule.actionResultReceived(action2);
+    buildSummaryStatsModule.buildComplete(createBuildEvent());
+    verify(reporterMock).handle(Event.info("CPU time ???s (user ???s, system ???s, bazel jvm ???s)"));
+  }
+
+  @Test
+  public void knownAndUnknownCpuTimesForAnActionIsReportedAndSumBecomeUnknown() throws Exception {
+    ActionResultReceivedEvent action1 = createActionEvent(Duration.ofSeconds(50), null);
+    buildSummaryStatsModule.actionResultReceived(action1);
+    buildSummaryStatsModule.buildComplete(createBuildEvent());
+    verify(reporterMock).handle(Event.info(String.format("CPU time ???s (user %.2fs, system ???s, bazel jvm ???s)", 50.00)));
+  }
+
+  @Test
+  public void reusedBuildSummaryStatsModuleIsClearedBetweenBuilds() throws Exception {
+    ActionResultReceivedEvent action1 = createActionEvent(Duration.ofSeconds(50),
+                                                          Duration.ofSeconds(20));
+    buildSummaryStatsModule.actionResultReceived(action1);
+    Field field1 = tBSSM.getDeclaredField("cpuTimeForBazelJvm");
```

Comment:
Please create `@VisibleOnlyForTesting` method with setter of this field, and use this method instead of using this mechanism

Reply:
Same for other places, where you are using this mechanism.